### PR TITLE
Eyecandy update

### PR DESCRIPTION
**Pull request by rafern**
grab some coffee this ones got a lot of code (sorry)
- fixed ambiance module killing you if you have epilepsy
  - it was unnecessarily binding textures causing it to flash a lot in color mode
- re-did most of the crosshair module:
  - added crosshair thickness, outline color, outline thickness and fill invert option
  - fill invert is basically how vanilla renders crosshair by inverting the crosshair's background so its always visible
  - fixed the alpha value not being updated when changed
- added shader system
  - new shaderprogram class, replaces the old shader class which was removed
  - shaderprogram can be loaded from json files or created in code
  - can load from both filesystem and resource paths, but resource paths must have a `resource://` prefix
  - automatically sets needed uniforms, even user created ones (but for now, user created uniforms only use the default values)
  - properly cleans up; you can nest shaders for example. it does this by keeping track of the used shaders in a stack. this also makes it compatible with optifine (with some caviats, next point)
  - you can get the depth of the screen by using the framebufferutil class (this doesnt work for optifine if you are using optifine shaders, the depth will always be zero if you have an optifine shader enabled, i dont know how to fix this yet, still need to reverse engineer how optifine does depth textures)
  - shaders are only compiled when needed
  - works with .reload
  - all shaders are tracked in a shadermanager class (also found in the seppuku class)
  - contains a utility class for setting most uniform types
  - you can use shaders as module values by using the shader class (not shaderprogram) as the value's type, they will show up in the ui as a carousel component. also settable via commands
  - has some default shaders (default 3d shader; mimics minecraft rendering, chams 3d shader; see next point)
- updated chams module
  - added shader mode; uses new shader system to render entities. the default shader for this looks like this:
shader_demo-min
  - added color options for unknown entity types (disabled by default)
  - chams uses a higher polygon offset; if you stood against a wall, chams would fail. this fixes it
  - cleaned up the code
  - shadows and fire are properly rendered now; before, you would never get shadows again after enabling chams

quirks:
- minecraft renders entity brightness (primed tnt glow, mobs going red when hurt) by combining textures, which isnt supported by shaders. to get around this, 2 hooks were added to keep track of the entity brightness value. the default shaders try to mimic entity brightness, but it doesnt look the same. it looks good though so i kept it

TODO
- shaders are loaded twice by the shadermanager. dont know why but it doesnt have an effect since they just get destroyed and reloaded. will fix in the future
- no way to set user uniforms yet; for now only default values are used, but the system is there (min, max, type, etc...). im burned out, but i will update the ui system in the future to be more modular since the ui for this would basically be the module list ui, but with shaders instead of modules and uniforms instead of module values
- shaders for skyboxes
- i have no idea how to fix name rendering with shaders, but will look for a fix in the future (maybe render them separately):
(no shader)
no-shader-name
(shader)
shader-name

### CODE REVIEW

File: src/main/java/me/rigamortis/seppuku/impl/module/render/ChamsModule.java [hunk 1]
```diff
@@ -69,148 +114,225 @@ public String getMetaData() {
...
-                        glEnable(GL11.GL_LIGHTING);
+                        GlStateManager.disablePolygonOffset();
                         break;
                     case "wireframe":
-                        glPolygonMode(GL11.GL_FRONT_AND_BACK, GL11.GL_FILL);
-                        glDisable(GL11.GL_POLYGON_OFFSET_LINE);
-                        glPolygonOffset(1.0f, 1100000.0f);
-                        glEnable(GL11.GL_TEXTURE_2D);
-                        glEnable(GL11.GL_LIGHTING);
+                        // GlStateManager.polygonOffsetState.polygonOffsetLine is private :(
```
Comment (uoil) [inline src/main/java/me/rigamortis/seppuku/impl/module/render/ChamsModule.java:185]: wonder if access transformer make this public

**Review by uoil — APPROVED** [review]
gotta check them out still in game but this is looking great dude

**Comment by rafern** [comment]
@uoil you want me to start doing smaller commits or is it okay to just keep doing big commits like these?

**Comment by uoil** [comment]
however your workflow is, is perfectly fine, if you know you'll have to edit some of the API maybe send those updates smaller i don't mind having to do manual merging if it happens

**Comment by uoil** [comment]
also i think turning on shaders in chams brings out a bug in block highlight where the render is glitched, turning on Storage will fix it so i think it's just states not being correct in block highlight

**Comment by rafern** [comment]
thats weird, ill see if i can fix it tomorrow

**Comment by Old-Chum** [comment]
> i have no idea how to fix name rendering with shaders, but will look for a fix in the future (maybe render them separately):
(no shader)

You can get around this by making an event for `EntityRenderer#drawNameplate` and cancelling it while you are rendering the entities with a shader, then re-drawing all the cancelled nametags after you are done with the shader.

**Pull request merged**